### PR TITLE
fix: adjust rcmgr limits for accelerated DHT client rt refresh

### PR DESCRIPTION
**Pull request by guseggert**
The Accelerated DHT client periodically refreshes its routing table,
including at startup, and if Resource Manager throttling causes the
client's routing table to be incomplete, then content routing may be
degraded or broken for users.

This adjusts the default limits to a level that empirically doesn't
cause Resource Manager throttling during initial DHT client
bootstrapping. Ideally the Accelerated DHT client would handle this
scenario more gracefully, but this works for now to unblock the 0.13
release.

### CODE REVIEW

File: core/node/libp2p/rcmgr_defaults.go [hunk 1]
```diff
@@ -39,9 +39,10 @@ func adjustedDefaultLimits(cfg config.SwarmConfig) rcmgr.DefaultLimitConfig {
 		maxconns := cfg.ConnMgr.HighWater
 		if 2*maxconns > defaultLimits.SystemBaseLimit.ConnsInbound {
 			// adjust conns to 2x to allow for two conns per peer (TCP+QUIC)
-			defaultLimits.SystemBaseLimit.ConnsInbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.ConnsOutbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.Conns = logScale(4 * maxconns)
+			// and 2x again to allow for burst of conns during accelerated DHT client rt refresh
```
Comment (aschmahmann) [inline core/node/libp2p/rcmgr_defaults.go:42]: This PR should include https://github.com/libp2p/go-libp2p-kad-dht/pull/772/. As a way of detecting errors.

For example, by bumping go-cid in that PR there's some (relatively small) refactoring you'll likely need to do in this PR to deal with the relatively small break in go-cid https://github.com/ipfs/go-cid/releases/tag/v0.2.0

File: core/node/libp2p/rcmgr_defaults.go [hunk 2]
```diff
@@ -39,9 +39,10 @@ func adjustedDefaultLimits(cfg config.SwarmConfig) rcmgr.DefaultLimitConfig {
...
+			defaultLimits.SystemBaseLimit.ConnsInbound = logScale(4 * maxconns)
+			defaultLimits.SystemBaseLimit.ConnsOutbound = logScale(4 * maxconns)
+			defaultLimits.SystemBaseLimit.Conns = logScale(8 * maxconns)
```
Comment (aschmahmann) [inline core/node/libp2p/rcmgr_defaults.go:45]: I'm not sure that only doing this in the context of  `2*HighWater > defaultLimits.SystemBaseLimit.ConnsInbound` makes sense. Even if your HighWater is lower (e.g. 100 connections) you might still want to allow for spikiness when your application demands it (e.g. for the accelerated DHT client routing table refreshes).

**Comment by BigLep** [comment]
2022-05-19 conversation: 
1. Set outbound limits really high (e.g., 64k) (we're not touching inbound limits)
2. @guseggert  Smoke test that we no longer get the error
We're not going to do wider scenario testing.

Pros:
This lets us ship
We at least get DoS protection on the inbound side

File: core/node/libp2p/rcmgr_defaults.go (hunk 1 above)
Comment (guseggert) [reply core/node/libp2p/rcmgr_defaults.go]: yeah that's fine I was going to follow up with a PR for that

File: core/node/libp2p/rcmgr_defaults.go (hunk 2 above)
Comment (guseggert) [reply core/node/libp2p/rcmgr_defaults.go]: Do you have a suggestion here?  I'm mostly focused on making sure this works for the default scenario, not for folks tweaking connmgr or rcmgr settings.

File: core/node/libp2p/rcmgr_defaults.go [hunk 3]
```diff
@@ -38,14 +38,17 @@ func adjustedDefaultLimits(cfg config.SwarmConfig) rcmgr.DefaultLimitConfig {
...
 		if 2*maxconns > defaultLimits.SystemBaseLimit.ConnsInbound {
-			// adjust conns to 2x to allow for two conns per peer (TCP+QUIC)
+			// Conns should be at least 2x larger than the high water to allow for two conns per peer (TCP+QUIC).
+			//
+			// Outbound conns are set very high to allow for the accelerated DHT client to refresh its routing table.
+			// Currently it does not gracefully handle RM throttling, once it does we can lower this.
 			defaultLimits.SystemBaseLimit.ConnsInbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.ConnsOutbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.Conns = logScale(4 * maxconns)
+			defaultLimits.SystemBaseLimit.ConnsOutbound = logScale(64 * maxconns)
```
Comment (aschmahmann) [inline core/node/libp2p/rcmgr_defaults.go:46]: This doesn't seem sufficient to effectively turn off outbound limits. Local debugging still shows errors due to a cap of ~2k connections, I'm not sure which limit this is coming from though.

**Comment by BigLep** [comment]
I saw this comment coment come through:

> Local debugging still shows errors due to a cap of ~2k connections, I'm not sure which limit this is coming from though.

I'm concerned about the "I'm not sure which limit this is coming from though".  Is this made clear in the logs?

File: core/node/libp2p/rcmgr_defaults.go [hunk 4]
```diff
@@ -25,32 +25,38 @@ func adjustedDefaultLimits(cfg config.SwarmConfig) rcmgr.DefaultLimitConfig {
...
+			// Conns should be at least 2x larger than the high water to allow for two conns per peer (TCP+QUIC).
 			defaultLimits.SystemBaseLimit.ConnsInbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.ConnsOutbound = logScale(2 * maxconns)
-			defaultLimits.SystemBaseLimit.Conns = logScale(4 * maxconns)
 
-			defaultLimits.SystemBaseLimit.StreamsInbound = logScale(16 * maxconns)
-			defaultLimits.SystemBaseLimit.StreamsOutbound = logScale(64 * maxconns)
-			defaultLimits.SystemBaseLimit.Streams = logScale(64 * maxconns)
+			// We want the floor of outbound conns to be no less than what was set above.
+			if outbound := logScale(64 * maxconns); outbound > defaultLimits.SystemBaseLimit.ConnsOutbound {
```
Comment (aschmahmann) [inline core/node/libp2p/rcmgr_defaults.go:48]: Did you mean for this to be 64 or just 2 (i.e. as it was before this PR)?

**Review by aschmahmann — APPROVED** [review]

File: core/node/libp2p/rcmgr_defaults.go [hunk 5]
```diff
@@ -25,32 +25,38 @@ func adjustedDefaultLimits(cfg config.SwarmConfig) rcmgr.DefaultLimitConfig {
...
-			defaultLimits.SystemBaseLimit.StreamsOutbound = logScale(64 * maxconns)
-			defaultLimits.SystemBaseLimit.Streams = logScale(64 * maxconns)
+			// We want the floor of outbound conns to be no less than what was set above.
+			if outbound := logScale(2 * maxconns); outbound > defaultLimits.SystemBaseLimit.ConnsOutbound {
+				defaultLimits.SystemBaseLimit.ConnsOutbound = outbound
+			}
 
 			if 2*maxconns > defaultLimits.SystemBaseLimit.FD {
 				defaultLimits.SystemBaseLimit.FD = logScale(2 * maxconns)
 			}
```
Comment (aschmahmann) [inline core/node/libp2p/rcmgr_defaults.go:54]: More FDs might be needed here (perhaps we need a default floor). I ran into this issue locally with lowered FDs. To be fair though we're leaving RCMgr off by default until we get better debugging and hopefully the go-libp2p-kad-dht changes will be enough to prompt people who lower their peer limits to increase their FD limits if they turn RCMgr on

**Pull request merged**